// bson/src/parser/serializer.ts
ey, value, index);
      } else if (value === null) {
        index = serializeNull(buffer, key, value, index);
      } else if (type === 'string') {
        index = serializeString(buffer, key, value, index);
      } else if (type === 'number') {
        index = serializeNumber(buffer, key, value, index);
      } else if (type === 'bigint') {
        index = serializeBigInt(buffer, key, value, index);
      } else if (type === 'boolean') {
        index = serializeBoolean(buffer, key, value, index);
      } else if (type === 'object' && value._bsontype == null) {
        if (value instanceof Date || isDate(value)) {
          index = serializeDate(buffer, key, value, index);
        } else if (value instanceof Uint8Array || isUint8Array(value)) {
          index = serializeBuffer(buffer, key, value, index);
        } else if (value instanceof RegExp || isRegExp(value)) {
          index = serializeRegExp(buffer, key, value, index);
        } else {
          index = serializeObject(
            buffer,
            key,
            value,
            index,
            checkKeys,
            depth,
            serializeFunctions,
            ignoreUndefined,
            path
          );
        }
      } else if (type === 'object') {
        if (value[constants.BSON_VERSION_SYMBOL] !== constants.BSON_MAJOR_VERSION) {
          throw new BSONVersionError();
        } else if (value._bsontype === 'ObjectId') {
          index = serializeObjectId(buffer, key, value, index);
        } else if (value._bsontype === 'Decimal128') {
          index = serializeDecimal128(buffer, key, value, index);
        } else if (value._bsontype === 'Long' || value._bsontype === 'Timestamp') {
          index = serializeLong(buffer, key, value, index);
        } else if (value._bsontype === 'Double') {
          index = serializeDouble(buffer, key, value, index);
        } else if (value._bsontype === 'Code') {
          index = serializeCode(
            buffer,
            key,
            value,
            index,
            checkKeys,
            depth,
            serializeFunctions,
            ignoreUndefined,
            path
          );
        } else if (value._bsontype === 'Binary') {
          index = serializeBinary(buffer, key, value, index);
        } else if (value._bsontype === 'BSONSymbol') {
          index = serializeSymbol(buffer, key, value, index);
        } else if (value._bsontype === 'DBRef') {
          index = serializeDBRef(buffer, key, value, index, depth, serializeFunctions, path);
        } else if (value._bsontype === 'BSONRegExp') {
          index = serializeBSONRegExp(buffer, key, value, index);
        } else if (value._bsontype === 'Int32') {
          index = serializeInt32(buffer, key, value, index);
        } else if (value._bsontype === 'MinKey' || value._bsontype === 'MaxKey') {
          index = serializeMinMax(buffer, key, value, index);
        } else if (typeof value._bsontype !== 'undefined') {
          throw new BSONError(`Unrecognized or invalid _bsontype: ${String(value._bsontype)}`);
        }
      } else if (type === 'function' && serializeFunctions) {
        index = serializeFunction(buffer, key, value, index);
      }
    }
  } else if (object instanceof Map || isMap(object)) {
    const iterator = object.entries();
    let done = false;

    while (!done) {
      // Unpack the next entry
      const entry = iterator.next();
      done = !!entry.done;
      // Are we done, then skip and terminate
      if (done) continue;

      // Get the entry values
      const key = entry.value ? entry.value[0] : undefined;
      let value = entry.value ? entry.value[1] : undefined;

      if (typeof value?.toBSON === 'function') {
        value = value.toBSON();
      }

      // Check the type of the value
      const type = typeof value;

      // Check the key and throw error if it's illegal
      if (typeof key === 'string' && !ignoreKeys.has(key)) {
        if (key.match(regexp) != null) {
          // The BSON spec doesn't allow keys with null bytes because keys are
          // null-terminated.
          throw new BSONError('key ' + key + ' must not contain null bytes');
        }

        if (checkKeys) {
          if ('$' === key[0]) {
            throw new BSONError('key ' + key + " must not start with '$'");
          } else if (key.includes('.')) {
            throw new BSONError('key ' + key + " must not contain '.'");
          }
        }
      }

      if (value === undefined) {
        if (ignoreUndefined === false) index = serializeNull(buffer, key, value, index);
      } else if (value === null) {
        index = serializeNull(buffer, key, value, index);
      } else if (type === 'string') {
        index = serializeString(buffer, key, value, index);
      } else if (type === 'number') {
        index = serializeNumber(buffer, key, value, index);
      } else if (type === 'bigint') {
        index = serializeBigInt(buffer, key, value, index);
      } else if (type === 'boolean') {
        index = serializeBoolean(buffer, key, value, index);
      } else if (type === 'object' && value._bsontype == null) {
        if (value instanceof Date || isDate(value)) {
          index = serializeDate(buffer, key, value, index);
        } else if (value instanceof Uint8Array || isUint8Array(value)) {
          index = serializeBuffer(buffer, key, value, index);
        } else if (value instanceof RegExp || isRegExp(value)) {
          index = serializeRegExp(buffer, key, value, index);
        } else {
          index = serializeObject(
            buffer,
            key,
            value,
            index,
            checkKeys,
            depth,
            serializeFunctions,
            ignoreUndefined,
            path
          );
        }
      } else if (type === 'object') {
        if (value[constants.BSON_VERSION_SYMBOL] !== constants.BSON_MAJOR_VERSION) {
          throw new BSONVersionError();
        } else if (value._bsontype === 'ObjectId') {
          index = serializeObjectId(buffer, key, value, index);
        } else if (value._bsontype === 'Decimal128') {
          index = serializeDecimal128(buffer, key, value, index);
        } else if (value._bsontype === 'Long' || value._bsontype === 'Timestamp') {
          index = serializeLong(buffer, key, value, index);
        } else if (value._bsontype === 'Double') {
          index = serializeDouble(buffer, key, value, index);
        } else if (value._bsontype === 'Code') {
          index = serializeCode(
            buffer,
            key,
            value,
            index,
            checkKeys,
            depth,
            serializeFunctions,
            ignoreUndefined,
            path
          );
        } else if (value._bsontype === 'Binary') {
          index = serializeBinary(buffer, key, value, index);
        } else if (value._bsontype === 'BSONSymbol') {
          index = serializeSymbol(buffer, key, value, index);
        } else if (value._bsontype === 'DBRef') {
          index = serializeDBRef(buffer, key, value, index, depth, serializeFunctions, path);
        } else if (value._bsontype === 'BSONRegExp') {
          index = serializeBSONRegExp(buffer, key, value, index);
        } else if (value._bsontype === 'Int32') {
          index = serializeInt32(buffer, key, value, index);
        } else if (value._bsontype === 'MinKey' || value._bsontype === 'MaxKey') {
          index = serializeMinMax(buffer, key, value, index);
        } else if (typeof value._bsontype !== 'undefined') {
          throw new BSONError(`Unrecognized or invalid _bsontype: ${String(value._bsontype)}`);
        }
      } else if (type === 'function' && serializeFunctions) {
        index = serializeFunction(buffer, key, value, index);
      }
    }
  } else {
    if (typeof object?.toBSON === 'function') {
      // Provided a custom serialization method
      object = object.toBSON();
      if (object != null && typeof object !== 'object') {
        throw new BSONError('toBSON function did not return an object');
      }
    }

    // Iterate over all the keys
    for (const key of Object.keys(object)) {
      let value = object[key];
      // Is there an override value
      if (typeof value?.toBSON === 'function') {
        value = value.toBSON();
      }

      // Check the type of the value
      const type = typeof value;

      // Check the key and throw error if it's illegal
      if (typeof key === 'string' && !ignoreKeys.has(key)) {
        if (key.match(regexp) != null) {
          // The BSON spec doesn't allow keys with null bytes because keys are
          // null-terminated.
          throw new BSONError('key ' + key + ' must not contain null bytes');
        }

        if (checkKeys) {
          if ('$' === key[0]) {
            throw new BSONError('key ' + key + " must not start with '$'");
          } else if (key.includes('.')) {
            throw new BSONError('key ' + key + " must not contain '.'");
          }
        }
      }

      if (value === undefined) {
        if (ignoreUndefined === false) index = serializeNull(buffer, key, value, index);
      } else if (value === null) {
        index = serializeNull(buffer, key, value, index);
      } else if (type === 'string') {
        index = serializeString(buffer, key, value, index);
      } else if (type === 'number') {
        index = serializeNumber(buffer, key, value, index);
      } else if (type === 'bigint') {
        index = serializeBigInt(buffer, key, value, index);
      } else if (type === 'boolean') {
        index = serializeBoolean(buffer, key, value, index);
      } else if (type === 'object' && value._bsontype == null) {
        if (value instanceof Date || isDate(value)) {
          index = serializeDate(buffer, key, value, index);
        } else if (value instanceof Uint8Array || isUint8Array(value)) {
          index = serializeBuffer(buffer, key, value, index);
        } else if (value instanceof RegExp || isRegExp(value)) {
          index = serializeRegExp(buffer, key, value, index);
        } else {
          index = serializeObject(
            buffer,
            key,
            value,
            index,
            checkKeys,
            depth,
            serializeFunctions,
            ignoreUndefined,
            path
          );
        }
      } else if (type === 'object') {
        if (value[constants.BSON_VERSION_SYMBOL] !== constants.BSON_MAJOR_VERSION) {
          throw new BSONVersionError();
        } else if (value._bsontype === 'ObjectId') {
          index = serializeObjectId(buffer, key, value, index);
        } else if (value._bsontype === 'Decimal128') {
          index = serializeDecimal128(buffer, key, value, index);
        } else if (value._bsontype === 'Long' || value._bsontype === 'Timestamp') {
          index = serializeLong(buffer, key, value, index);
        } else if (value._bsontype === 'Double') {
          index = serializeDouble(buffer, key, value, index);
        } else if (value._bsontype === 'Code') {
          index = serializeCode(
            buffer,
            key,
            value,
            index,
            checkKeys,
            depth,
            serializeFunctions,
            ignoreUndefined,
            path
          );
        } else if (value._bsontype === 'Binary') {
          index = serializeBinary(buffer, key, value, index);
        } else if (value._bsontype === 'BSONSymbol') {
          index = serializeSymbol(buffer, key, value, index);
        } else if (value._bsontype === 'DBRef') {
          index = serializeDBRef(buffer, key, value, index, depth, serializeFunctions, path);
        } else if (value._bsontype === 'BSONRegExp') {
          index = serializeBSONRegExp(buffer, key, value, index);
        } else if (value._bsontype === 'Int32') {
          index = serializeInt32(buffer, key, value, index);
        } else if (value._bsontype === 'MinKey' || value._bsontype === 'MaxKey') {
          index = serializeMinMax(buffer, key, value, index);
        } else if (typeof value._bsontype !== 'undefined') {
          throw new BSONError(`Unrecognized or invalid _bsontype: ${String(value._bsontype)}`);
        }
      } else if (type === 'function' && serializeFunctions) {
        index = serializeFunction(buffer, key, value, index);
      }
    }
  }

  // Remove the path
  path.delete(object);

  // Final padding byte for object
  buffer[index++] = 0x00;

  // Final size
  const size = index - startingIndex;
  // Write the size of the object
  startingIndex += NumberUtils.setInt32LE(buffer, startingIndex, size);
  return index;
}
ize <= 16) {
    for (let i = 0; i < size; i++) buffer[index + i] = data[i];
  } else {
    buffer.set(data, index);
  }
  // Adjust the index
  index = index + value.position;
  return index;
}

function serializeSymbol(buffer: Uint8Array, key: string, value: BSONSymbol, index: number) {
  // Write the type
  buffer[index++] = constants.BSON_DATA_SYMBOL;
  // Number of written bytes
  const numberOfWrittenBytes = ByteUtils.encodeUTF8Into(buffer, key, index);
  // Encode the name
  index = index + numberOfWrittenBytes;
  buffer[index++] = 0;
  // Write the string
  const size = ByteUtils.encodeUTF8Into(buffer, value.value, index + 4) + 1;
  // Write the size of the string to buffer
  NumberUtils.setInt32LE(buffer, index, size);
  // Update index
  index = index + 4 + size - 1;
  // Write zero
  buffer[index++] = 0;
  return index;
}

function serializeDBRef(
  buffer: Uint8Array,
  key: string,
  value: DBRef,
  index: number,
  depth: number,
  serializeFunctions: boolean,
  path: Set<Document>
) {
  // Write the type
  buffer[index++] = constants.BSON_DATA_OBJECT;
  // Number of written bytes
  const numberOfWrittenBytes = ByteUtils.encodeUTF8Into(buffer, key, index);

  // Encode the name
  index = index + numberOfWrittenBytes;
  buffer[index++] = 0;

  let startIndex = index;
  let output: DBRefLike = {
    $ref: value.collection || value.namespace, // "namespace" was what library 1.x called "collection"
    $id: value.oid
  };

  if (value.db != null) {
    output.$db = value.db;
  }

  output = Object.assign(output, value.fields);
  const endIndex = serializeInto(
    buffer,
    output,
    false,
    index,
    depth + 1,
    serializeFunctions,
    true,
    path
  );

  // Calculate object size
  const size = endIndex - startIndex;
  // Write the size
  startIndex += NumberUtils.setInt32LE(buffer, index, size);
  // Set index
  return endIndex;
}

export function serializeInto(
  buffer: Uint8Array,
  object: Document,
  checkKeys: boolean,
  startingIndex: number,
  depth: number,
  serializeFunctions: boolean,
  ignoreUndefined: boolean,
  path: Set<Document> | null
): number {
  if (path == null) {
    // We are at the root input
    if (object == null) {
      // ONLY the root should turn into an empty document
      // BSON Empty document has a size of 5 (LE)
      buffer[0] = 0x05;
      buffer[1] = 0x00;
      buffer[2] = 0x00;
      buffer[3] = 0x00;
      // All documents end with null terminator
      buffer[4] = 0x00;
      return 5;
    }

    if (Array.isArray(object)) {
      throw new BSONError('serialize does not support an array as the root input');
    }
    if (typeof object !== 'object') {
      throw new BSONError('serialize does not support non-object as the root input');
    } else if ('_bsontype' in object && typeof object._bsontype === 'string') {
      throw new BSONError(`BSON types cannot be serialized as a document`);
    } else if (
      isDate(object) ||
      isRegExp(object) ||
      isUint8Array(object) ||
      isAnyArrayBuffer(object)
    ) {
      throw new BSONError(`date, regexp, typedarray, and arraybuffer cannot be BSON documents`);
    }

    path = new Set();
  }

  // Push the object to the path
  path.add(object);

  // Start place to serialize into
  let index = startingIndex + 4;

  // Special case isArray
  if (Array.isArray(object)) {
    // Get object keys
    for (let i = 0; i < object.length; i++) {
      const key = `${i}`;
      let value = object[i];

      // Is there an override value
      if (typeof value?.toBSON === 'function') {
        value = value.toBSON();
      }

      // Check the type of the value
      const type = typeof value;

      if (value === undefined) {
        index = serializeNull(buffer, k
